src/firefox_client.ts

```diff
@@ -0,0 +1,658 @@
+   *
+   * @returns The tab, holding the actor we use every other request
+   */
+  private async listTabs(): Promise<Tab> {
+    let listTabsResponse = await this.request("listTabs", {}, "root");
+    // NOTE: When browser isn't ran in headless, there is usually 2 tabs open, the first one being "Advanced Preferences" or "about" page, and the second one being the actual page we navigated to
+    let tabs = listTabsResponse.tabs;
+    while (
+      tabs.length === 0 || (tabs.length > 0 && tabs[0].title === "New Tab")
+    ) {
+      listTabsResponse = await this.request("listTabs", {}, "root");
+      tabs = listTabsResponse.tabs;
+    }
+    let tab = listTabsResponse.tabs.find((t: Tab) =>
+      t.selected === true
+    ) as Tab;
+    // For firefox > 75 consoleActor is not available within listTabs request
+    if (tab && !tab.consoleActor) {
+      const tabActorRequest = await this.request("getTarget", {}, tab.actor);
+      tab = {
+        ...tab,
+        ...tabActorRequest.frame,
+      };
+    }
+    return tab;
+  }
+
+  /**
+   * The main method we use to send requests through the connection we
+   * establish with the headless browser.
+   *
+   * @param type - The request type, eg "navigateTo"
+   * @param params - Any extra parameters the request will use. For example, if using `navigateTo`: `await this.request("navigateTo", { url: "https://google.com })`
+   * @param actor - As opposed to using the actor for the tab, you can override it with an actor of your choice. This is mainly here so we can use "root" as the actor when listing tabs (see `listTabs`)
+   *
+   * @returns The response of the request.
+   * An object containing the message: This is a parsed JSON response that was assigned to th id, thus the response from thee request
+   */
+  private async request(
+    type: string,
+    params = {},
+    actor?: string,
+  ): Promise<Packet> {
+    actor = actor ? actor : this.tab!.actor;
+    // Construct data in required format to send
+    const message = {
+      type,
+      to: actor,
+      ...params,
+    };
+    const str = JSON.stringify(message);
+    const encodedMessage = `${(encoder.encode(str)).length}:${str}`;
+    // Send message
+    await this.conn.write(encoder.encode(encodedMessage));
+    // Get related packet
+    const packet = await this.waitForSpecificPacket(actor, {
+      from: actor,
+    });
+    // Check for errors
+    if ("error" in packet) {
+      await this.done(`${packet.error}: ${packet.message}`);
+    }
+    if ("pageError" in packet) {
+      await this.done(`${packet.type}: ${packet.errrorMessage}`);
+    }
+    // Return result
+    return packet;
+  }
+
+  /**
+   * Get full path to the firefox binary on the user'ss filesystem.
+   * Thanks to [caspervonb](https://github.com/caspervonb/deno-web/blob/master/browser.ts)
+   *
+   * @returns the path
+   */
+  private static getFirefoxPath(): string {
```

Comment:
what if the machine doesn't have FF installed?

Reply:
It’s going to throw a file not found error when it tries to run Firefox from the sub process